### PR TITLE
Add a documentation tokens to pps

### LINKED ISSUE #266: gospel pps detaches documentation comments when below and followed by a specification
This behaviour can be reproduced as follow:

```
$ cat > ./bug_pps.mli << EOF
> val f : int -> int
> (** blah *)
> (*@ y = f x *)
> EOF
```
 Then we have:
```
$ ocamlc -dsource bug_pps.mli
val f : int -> int[@@ocaml.doc " blah "]
```
Which is the expected behaviour.

But:
```
$ ocamlc -pp "gospel pps" -dsource bug_pps.mli
val f : int -> int[@@gospel {| y = f x |}]
```
Where the documentation comment has been lost.

### PR DESCRIPTION
**Pull request by n-osborne**
This PR proposes a fix for #288

It adds two new tokens to the lexer:
- a `Documentation` token with starting and ending position and the contents
- an `Empty_documentation` token with the ending position

Both are printed as documentation attributes when they appears between an Ocaml declaration and a gospel specification (the case where the documentation comments is considered as detached by the OCaml parser)

Tests are added to check that the documentation is not considered as detached anymore and to make sure that original position is preserved.

### CODE REVIEW

File: src/pps.mll [hunk 1]
```diff
@@ -151,6 +170,19 @@ and gospel start_pos = parse
       scan lexbuf
     }
 
+and documentation start_pos = parse
+  | blank+ as s { Buffer.add_string buf s; documentation start_pos lexbuf }
+  | newline as nl { Buffer.add_string buf nl; Lexing.new_line lexbuf; documentation start_pos lexbuf }
+  | "*)" {
```
Comment (shym) [inline src/pps.mll:176]: This doesn’t support nested comments, but it would be nice to make sure they are, using a similar trick to was is done for regular comments.

File: test/documentation/test_preprocess_documentation.t [hunk 2]
```diff
@@ -0,0 +1,65 @@
+Testing interaction between `gospel pps` and Ocaml compiler. The issue is that
```
Comment (shym) [inline test/documentation/test_preprocess_documentation.t:1]: Very nice test!
Doesn’t it miss some dune configuration to be run? I intentionally broke the test and `dune runtest` didn’t complain.

File: src/pps.mll [hunk 3]
```diff
@@ -82,8 +95,12 @@
     Fmt.str "%s%s" (print_gospel `TwoAt start_p end_p s) (print l)
   | Other o :: Spaces sp :: Spec (start_p, end_p, s) :: l ->
     Fmt.str "%s%s%s%s" o sp (print_gospel `TwoAt start_p end_p s) (print l)
+  | Other o :: Spaces sp :: Documentation (doc_start_p, doc_end_p, d) :: Spaces sp' :: Spec (start_p, end_p, s) :: l ->
```
Comment (shym) [inline src/pps.mll:98]: That would be admittedly rather ugly, but the `Spaces` mean that
```
val f: int -> int(** doc *)(*@ spec *)
```
isn’t covered as one would hope.

**Review by shym — COMMENTED** [review]
Very nice job!
I think I found two corner cases that it would be nice to cover before merging.

File: test/documentation/test_preprocess_documentation.t [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
```
Comment (shym) [inline test/documentation/test_preprocess_documentation.t:1]: The tests could also use an inner nested comment, since they should be now covered.

File: src/pps.mll [hunk 5]
```diff
@@ -120,6 +165,21 @@ rule scan = parse
     { push (); Lexing.new_line lexbuf; Queue.push (Spaces nl) queue; scan lexbuf }
   | "(*@"
     { push (); gospel (Lexing.lexeme_start_p lexbuf) lexbuf }
+  | "(**)" {
+    push ();
+    let end_pos = Lexing.lexeme_end_p lexbuf in
+    Queue.push (Empty_documentation end_pos) queue;
+    scan lexbuf }
```
Comment (shym) [inline src/pps.mll:172]: The two new rules use different block endings (`}` on a line by itself or not.

**Review by shym — APPROVED** [review]
Very nice job!
I managed to find two minor points just to quibble :-) but LGTM.
